builtin/array.mbt

```diff
@@ -577,6 +577,66 @@ pub fn search[T : Eq](self : Array[T], value : T) -> Int? {
   }
 }
 
+/// Binary search the array index for a given element in a sorted array.
+/// 
+/// If the element is found then Ok is returned, containing the matching array index. 
+/// If there are multiple matches, then the leftmost match could be returned. 
+/// If the element is not found then Err is returned, containing the array index where a matching element cloud be inserted while maintaining sorted order.
+/// 
+/// The array must be sorted in increaing order. If not, the returned result is meaningless.
+///
+/// # Example
+/// ```
+/// let v = [3, 4, 5]
+/// v.binary_search(3)
+/// ```
+pub fn binary_search[T : Compare](
+  self : Array[T],
+  value : T
+) -> Result[Int, Int] {
+  return self.binary_search_by(
+    fn(v : T) { if v < value { -1 } else if v == value { 0 } else { 1 } },
+  )
+}
+
+/// Binary search the array index for a given element in a sorted array with a custom comparator function.
```

Comment:
The wording here probably needs some improvement, here is an example from GPT4

```
/// Performs a binary search on a sorted array using a custom comparator function, and returns the index of the given element if found.
/// 
/// The comparator function should return an `Int` that indicates whether the argument is less than `(-1)`, equal to `(0)`, or greater than `(1)` 
/// the target element.
/// 
/// If the element is found, an `Ok` variant is returned containing the index of the matching element in the array. 
/// If there are multiple matches, the leftmost match will be returned. 
/// 
/// If the element is not found, an `Err` variant is returned containing the index where the element could be inserted while maintaining the sorted order of the array.
/// 
/// # Important Considerations
/// - The array must be sorted, and the comparator function should be consistent with the sorted order of the array.
/// - If the array is not sorted or if the comparator function is not consistent with the sorted order, the result of this function will be meaningless.
/// 
/// # Example
/// ```
/// let v = [3, 4, 5];
/// let result = v.binary_search_by(|v| {
///     if v < 3 { -1 }
///     else if v == 3 { 0 }
///     else { 1 }
/// });
/// assert_eq!(result, Ok(0)); // The element 3 is found at index 0
/// ```
/// 
/// # Type Parameters
/// - `T`: The type of elements in the array. It must implement the `Compare` trait.
/// 
/// # Arguments
/// - `self`: The array in which to perform the search.
/// - `cmp`: A closure that defines the comparator function, taking an element of type `T` and returning an `Int`.
/// 
/// # Returns
/// - `Result[Int, Int]`: If the element is found, `Ok` is returned with the index of the element.
/// If the element is not found, `Err` is returned with the index where the element could be inserted to maintain sorted order.
```

Reply:
I've no clue what the standard style for the document is. Your suggestions are inspiring. I'll adopt to this style.